e2e_tests/tests/experiment/test_custom_searcher.py

```diff
@@ -219,3 +225,314 @@ def sample_params(self) -> Dict[str, int]:
+            metric *= -1
+        return self.promote_async(request_id, metric)
+
+    def on_trial_exited_early(
+        self, request_id: uuid.UUID, exited_reason: ExitedReason
+    ) -> List[Operation]:
+        self.pending_trials -= 1
+        if (
+            exited_reason == ExitedReason.INVALID_HP
+            or exited_reason == ExitedReason.INIT_INVALID_HP
+        ):
+            ops: List[Operation] = []
+
+            self.early_exit_trials[request_id] = True
+            ops.append(Close(request_id))
+            self.closed_trials[request_id] = True
+            self.invalid_trials += 1
+
+            highest_rung_index = self.trial_rungs[request_id]
+            rung = self.rungs[highest_rung_index]
+            rung.outstanding_trials -= 1
+
+            for rung_idx in range(0, highest_rung_index + 1):
+                rung = self.rungs[rung_idx]
+                rung.metrics = list(filter(lambda x: x.request_id != request_id, rung.metrics))
+
+            create = Create(
+                request_id=uuid.uuid4(),
+                hparams=self.sample_params(),
+                checkpoint=None,
+            )
+            ops.append(create)
+            ops.append(
+                ValidateAfter(request_id=create.request_id, length=self.rungs[0].units_needed)
+            )
+
+            self.trial_rungs[create.request_id] = 0
+            self.pending_trials += 1
+
+            return ops
+
+        self.early_exit_trials[request_id] = True
+        self.closed_trials[request_id] = True
+        return self.promote_async(request_id, 100.0)
+
+    def initial_operations(self) -> List[Operation]:
+        ops: List[Operation] = []
+
+        if self.max_concurrent_trials > 0:
+            max_concurrent_trials = min(self.max_concurrent_trials, self.max_trials)
+        else:
+            max_concurrent_trials = max(
+                1, min(int(pow(self.divisor, self.num_rungs - 1)), self.max_trials)
+            )
+
+        for _ in range(0, max_concurrent_trials):
+            create = Create(
+                request_id=uuid.uuid4(),
+                hparams=self.sample_params(),
+                checkpoint=None,
+            )
+            ops.append(create)
+            ops.append(
+                ValidateAfter(request_id=create.request_id, length=self.rungs[0].units_needed)
+            )
+
+            self.trial_rungs[create.request_id] = 0
+            self.pending_trials += 1
+
+        return ops
+
+    def promote_async(self, request_id: uuid.UUID, metric: float) -> List[Operation]:
+        rung_idx = self.trial_rungs[request_id]
+        rung = self.rungs[rung_idx]
+        rung.outstanding_trials -= 1
+        added_train_workload = False
+
+        ops: List[Operation] = []
+
+        if rung_idx == self.num_rungs - 1:
+            rung.metrics.append(TrialMetric(request_id=request_id, metric=metric))
+
+            if request_id in self.early_exit_trials and self.early_exit_trials[request_id] is True:
```

Comment:
Line 206 in `asha.go` says
`if !s.EarlyExitTrials[requestID] {` which is the _opposite_ of your condition.

Also, you might consider using a set as a python equivalent of golang's `map[model.RequestID]bool`.

Reply:
Good catch. thanks!
Changing from dictionaries to sets.